lib/core/macros.nim

```diff
@@ -1611,3 +1611,43 @@ when defined(nimMacrosSizealignof):
     ## from a field of a type. Therefore it only requires one argument
     ## instead of two. Returns a negative value if the Nim compiler
     ## does not know the offset.
+
+proc splitDefinition*(def: NimNode): tuple[lhs: NimNode, rhs: NimNode] =
+  ## allows library constructs such as: `byRef: a2 = expr`
+  doAssert def.kind == nnkStmtList and def.len == 1
+  let def2 = def[0]
+  doAssert def2.kind == nnkAsgn
+  let lhs = def2[0]
+  let rhs = def2[1]
+  expectKind(lhs, nnkIdent)
+  return (lhs, rhs)
+
+macro byRef*(def: untyped): untyped =
+  ## Defines a ref alias for lvalue expressions. The expression is evaluated
+  ## only once, and any side effects will only be evaluated once, at declaration
+  ## time.
+  runnableExamples:
+    var count = 0
+    proc identity(a: int): auto =
+      block: count.inc; a
+    var x = @[1,2,3]
+    byRef: x1=x[identity(1)] # the ref-alias is evaluated only here
+    doAssert count == 1 
+    x1 += 10
+    doAssert type(x1) is int # use x1 just like a normal variable
+    doAssert x == @[1,12,3]
+    doAssert count == 1 # count has not changed
+
+  let (name, exp) = splitDefinition(def)
+  result = quote do:
+    let myAddr = addr `exp`
```

Comment:
Why not just use `unsafeAddr` for everything? You won't be able to modify `myAddr` anyway since this is a ``let`` AFAIK.

Reply:
> Why not just use unsafeAddr for everything?

seems unsafe to me; I'd like to use `byRef` in most cases and only `byPtr` when needed, for the same reasons as the split addr vs unsafeAddr; helps when auditing code

> You won't be able to modify myAddr anyway since this is a let AFAIK.

that's not my understanding, see the example I added to unittest:
```nim
block byPtrfBlock:
  type Foo = object
    x: string
  proc fun(a: Foo): auto =
    doAssert not compiles (block: byRef: x=a.x)
    byPtr: x=a.x
    x[0]='X'
  let foo = Foo(x: "asdf")
  fun(foo)
  doAssert foo.x == "Xsdf"
```